### ISSUE TITLE
[feature request] Ability to flush logs

### ISSUE THREAD

**Issue by ed-alertedh** [ess-dmsc/graylog-logger#33]
I was just debugging a startup issue in our program where it locks up. It was hard to figure out from the logs (configured to output to both Console and GELF) where it got to because several messages remain in the buffer and are never flushed out.

It would be nice if I could manually flush the logs out, or even better if the logging thread flushed them after a certain timeout. I would be happy to prepare a PR if you are interested but I would probably need a little guidance around the best way to add it to the current architecture.

**Comment by SkyToGround** [comment]
As I have not touched the code for quite a while now I had to study it for a while to determine what it is actually doing. In short: Flushing "kind of" works by printing messages from a queue until a "stop thread" message is received. This message is created when the destructor of a log handler is called. The exception is the Graylog-handler which will exit immediately upon the calling of its destructor.

There is also a `emptyQueue()` member function for all the handlers but it does not actually work for all of them.

In short: I am not happy about the current situation and I will fix it when I have the time (probably beginning of next year). If you want to help me getting it done earlier, feel free.

**Comment by ed-alertedh** [comment]
OK, I will see if I find some time to attempt this but otherwise happy to wait until you find some time for it!

**Comment by SkyToGround** [comment]
@ed-alertedh I am almost done with a branch that implements flushing behaviour as well as some other improvements (#34). Feel free to have a look at it and review it. I am hoping to have one of my colleagues review that branch within a week or two which means that this is the expected time-line for it getting merged with the master branch.

**Comment by ed-alertedh** [comment]
@SkyToGround Looks great! I had been using fmtlib already so that integration will be really nice. To be honest I didn't end up spending enough time on this to understand your implementation details, so I probably can't offer a useful review of the internals. But what I can try to do is build against this branch and test it out in our application some time in the next week or two.
